Change Localizer Interface

zachwaffle4 opened the pull request:
This change gives the Localizer interface methods `getPoseEstimate: Pose2d` and `getVelocityEstimate: PoseVelocity2d` (and updates all 4 current localizer implementations to match that). This allows for Localizers that update under the hood (ie SparkFun OTOS or coprocessors such as goBILDA's Pinpoint). I moved both DriveLocalizer classes out of their respective drivetrain classes and into a Localization package as well as adding an example SparkFunOTOSLocalizer option. 

The pose delta localization system is great for processing the localizer on the Rev Hubs, but since OTOS and Pinpoint directly return the measured position and velocity to create Localizer implementations for them with the current system you'd essentially have the `update` method return `newPose - oldPose` only for MecanumDrive and TankDrive to later say `pose = pose + localizer.update()` which is relatively illogical in my opinion.

rbrott commented:
Neat! My high-level thoughts are
* The new `Localizer` API should be 
   ```java
    public interface Localizer {
        Pose2d getPose();
        void setPose(Pose2d pose);
        PoseVelocity2d update();
    }
    ```
    There should only be one update per loop that updates the pose and computes local velocities at the same time. Having the localizer own the pose will also make the OTOS/Pinpoint implementations simpler and get rid of the "weird cursed behavior". 
* I don't want to include an OTOS or Pinpoint implementation right now since they aren't officially supported for tuning, and I'm not sure what the right split between RR FTC & quickstart code will be. I'm also not sure what the right way to handle faults is (for OTOS especially). 
* I like having the drive localizers in the drive classes themselves.

zachwaffle4 commented:
> There should only be one update per loop that updates the pose and computes local velocities at the same time.

This I agree with and I was wondering how I could avoid that with my new system. I was never a fan of having to call `updatePoseEstimate` in order to update the `pose` field but since that seems to be required in order to ensure pose is only updated once per loop I'm fine with keeping that.

I moved the `pose` property from the drive classes to the Localizer implementations (and then I added `getPose` and `setPose` methods to both drive classes that just invoke the localizer's methods of the same name); I also changed the FollowTrajectoryAction and TurnAction inner classes of both to reflect that (by having them call `getPose`).

The OTOS can be removed; it was mostly included as an example to show how my new system would work for a localizer that doesn't rely on pose deltas.

zachwaffle4 commented:
It then occured to me that removing the `pose` property would break backwards compatibility so I put it back and updated `getPose`, `setPose`, and `updatePoseEstimate` to match.

j5155 commented on TeamCode/src/main/res/xml/teamwebcamcalibrations.xml:
???

zachwaffle4 replied:
what?

j5155 replied:
Why did you modify this file?

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/tuning/TuningOpModes.java:
```diff
@@ -59,8 +59,8 @@ public static void register(OpModeManager manager) {
-                if (md.localizer instanceof MecanumDrive.DriveLocalizer) {
-                    MecanumDrive.DriveLocalizer dl = (MecanumDrive.DriveLocalizer) md.localizer;
+                if (md.localizer instanceof MecanumDrive.MecanumDriveLocalizer) {
+                    MecanumDrive.MecanumDriveLocalizer dl = (MecanumDrive.MecanumDriveLocalizer) md.localizer;
```
If these are subclasses of the drive class, why rename them?

zachwaffle4 replied:
this was an artifact of my moving of the drive localizers out to separate files; could be changed

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/tuning/TuningOpModes.java:
```diff
@@ -110,8 +110,8 @@ public static void register(OpModeManager manager) {
-                if (td.localizer instanceof TankDrive.DriveLocalizer) {
-                    TankDrive.DriveLocalizer dl = (TankDrive.DriveLocalizer) td.localizer;
+                if (td.localizer instanceof TankDrive.TankDriveLocalizer) {
+                    TankDrive.TankDriveLocalizer dl = (TankDrive.TankDriveLocalizer) td.localizer;
```
If these are subclasses of the drive class, why rename them?

zachwaffle4 replied:
this was an artifact of my moving of the drive localizers out to separate files; will be changed back

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/TwoDeadWheelLocalizer.java:
```diff
@@ -13,7 +15,6 @@
 import com.acmerobotics.roadrunner.ftc.PositionVelocityPair;
 import com.acmerobotics.roadrunner.ftc.RawEncoder;
 import com.qualcomm.robotcore.hardware.DcMotorEx;
-import com.qualcomm.robotcore.hardware.DcMotorSimple;
```
Why was this import removed? This is needed to reverse encoders

zachwaffle4 replied:
marked as unused

rbrott replied:
I agree with j5155; see the other thread on this import

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/TwoDeadWheelLocalizer.java:
```diff
@@ -1,7 +1,9 @@
-package org.firstinspires.ftc.teamcode;
+package org.firstinspires.ftc.teamcode.localization;
```
Why was this moved?

zachwaffle4 replied:
my goal initially was to have all the localizers in one package separate from the drive file bc that seemed more clear

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/Localizer.java:
```diff
@@ -0,0 +1,25 @@
+     * you must call update() to update the pose estimate.
+     * @return the Localizer's current pose
+     */
+    Pose2d getPose();
```
Would getCachedPose or something be better for this, to make it clear that you're not actually reading the pose from hardware?

zachwaffle4 replied:
perhaps? `pose` by itself was the name of the property in the robot drive classes so this feels like the logical extension of that

rbrott replied:
I think `getPose()`. (IMO the SDK is straying from convention by using the getter/setter nomenclature for costly I/O operations.)

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TankDrive.java:
```diff
@@ -495,4 +411,108 @@ public TrajectoryActionBuilder actionBuilder(Pose2d beginPose) {
         );
     }
+
+    public class TankDriveLocalizer implements Localizer {
```
Why is this renamed and moved?

zachwaffle4 replied:
this was an artifact of my moving of the drive localizers out to separate files; will be changed back

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -114,101 +113,14 @@ kinematics.new WheelVelConstraint(PARAMS.maxWheelVel),
 
     public final Localizer localizer;
     public Pose2d pose;
-
```
??

zachwaffle4 replied:
what?

rbrott replied:
Generally it's good to avoid changing whitespace in random places unrelated to your changes. I don't usually care too much about that, but any unneeded change in the quickstart has the potential to cause git conflicts for many downstream users.

j5155 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -34,13 +33,13 @@
 import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
 import com.qualcomm.robotcore.hardware.DcMotor;
 import com.qualcomm.robotcore.hardware.DcMotorEx;
-import com.qualcomm.robotcore.hardware.DcMotorSimple;
```
This is needed for reversing encoders and motors

rbrott replied:
Yes, I keep this unused import around to make things a little easier for teams.

zachwaffle4 replied:
I put it back :)

rbrott replied:
I still see it removed -- is Android Studio optimizing imports on the fly?

zachwaffle4 replied:
I think so, considering I explicitly typed the import out

j5155 commented:
> It then occured to me that removing the `pose` property would break backwards compatibility so I put it back and updated `getPose`, `setPose`, and `updatePoseEstimate` to match.

So the problem is that, in the past, the way of modifying the current pose was to change the pose variable directly in the drive class. 
As I see it there are three possible ways to handle this:
1. Actively drop backwards compatibility (remove the pose variable, ONLY have getters and setters); this would break legacy code but this is still my preferred solution
2. Allow legacy code to read the pose variable, but silently ignore any writes to it (this is what your current code does). I strongly dislike this solution as it will cause difficult to find bugs and silently break a lot of people's existing code without them noticing. 
3. Jankily check the pose variable every loop to see if something changed it, and write the new pose if so. This is the solution my fork uses, but I dislike it a lot; it can lead to time delays between the pose variable being updated and the updated pose being written. The only reason I use it is that it maintains 100% backwards compatibility.

j5155 commented:
> The OTOS can be removed; it was mostly included as an example
...so you should remove it from the PR then?

zachwaffle4 commented:
I removed OTOS from the PR; I kept it as a [gist](https://gist.github.com/zachwaffle4/7efd87d1ef5e97b289589aa2cd602eaa). 

> As I see it there are three possible ways to handle this

I would love to simply break backwards compatibility for the sake of cleanliness; instead I added the temporary checks so that every time `updatePoseEstimate` is called we make sure that the drive class `pose` property matches that of the localizer's (and then set the localizer `pose` to match). Maybe in RoadRunner 1.1 we can force people to use getPose and setPose instead :shrug:

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -114,101 +113,14 @@ kinematics.new WheelVelConstraint(PARAMS.maxWheelVel),
     private final DownsampledWriter driveCommandWriter = new DownsampledWriter("DRIVE_COMMAND", 50_000_000);
     private final DownsampledWriter mecanumCommandWriter = new DownsampledWriter("MECANUM_COMMAND", 50_000_000);
 
-    public class DriveLocalizer implements Localizer {
```
Can you put the drive localizer back here?

zachwaffle4 replied:
done :)

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/Localizer.java:
```diff
@@ -0,0 +1,25 @@
+    Pose2d getPose();
+
+    /**
+     * Updates the Localizer's position estimate.
```
Pose estimate is more accurate here than position estimate.

zachwaffle4 replied:
done :)

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/ThreeDeadWheelLocalizer.java:
I don't really like the `localization` package change. This PR isn't adding any new localizer classes, and it's a pity to have to move those files and update package declarations + imports. (I personally think subpackages are overused in Java. And don't get me started about single-class files...)

zachwaffle4 replied:
everything is back to the main package now (and yes it would be so much better if we could just put them all in one file; if only we could write the quickstart in Kotlin too)

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/ThreeDeadWheelLocalizer.java:
```diff
@@ -48,9 +50,28 @@ public ThreeDeadWheelLocalizer(HardwareMap hardwareMap, double inPerTick) {
     }
 
-    public Twist2dDual<Time> update() {
+    public Twist2dDual<Time> calculatePoseDelta() {
```
Can you fold this into the `update()` method?

zachwaffle4 replied:
done

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/TwoDeadWheelLocalizer.java:
```diff
@@ -58,9 +60,28 @@ public TwoDeadWheelLocalizer(HardwareMap hardwareMap, IMU imu, double inPerTick)
     }
 
-    public Twist2dDual<Time> update() {
+    public Twist2dDual<Time> calculatePoseDelta() {
```
ditto

zachwaffle4 replied:
done

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/localization/TwoDeadWheelLocalizer.java:
```diff
@@ -58,9 +60,28 @@ public TwoDeadWheelLocalizer(HardwareMap hardwareMap, IMU imu, double inPerTick)
     }
 
-    public Twist2dDual<Time> update() {
+    public Twist2dDual<Time> calculatePoseDelta() {
```
ditto

rbrott replied:
ditto

rbrott reviewed (commented):
> I would love to simply break backwards compatibility for the sake of cleanliness; instead I added the temporary checks so that every time updatePoseEstimate is called we make sure that the drive class pose property matches that of the localizer's (and then set the localizer pose to match). Maybe in RoadRunner 1.1 we can force people to use getPose and setPose instead 🤷

I think `drive.pose` should be removed and users should use `drive.localizer.getPose()`/`drive.localizer.setPose()` (I don't like the extra indirection of getters and setters that simply forward to another type.) The docs can be updated to mention both versions.

zachwaffle4 commented:
> I think `drive.pose` should be removed and users should use `drive.localizer.getPose()`/`drive.localizer.setPose()`.

I did this (I added a `getLocalizerPose` method to the drive classes too, but the drive class no longer stores a pose independently of the localizers), but doing this would cause a lot of issues with people updating. no?

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -129,21 +125,36 @@ public class DriveLocalizer implements Localizer {
             rightFront = new OverflowEncoder(new RawEncoder(MecanumDrive.this.rightFront));
 
-            imu = lazyImu.get();
+            imu = MecanumDrive.this.lazyImu.get();
```
Can you revert this line?

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -166,14 +177,17 @@ public Twist2dDual<Time> update() {
                 lastHeading = heading;
 
-                return new Twist2dDual<>(
+                delta = new Twist2dDual<>(
                         Vector2dDual.constant(new Vector2d(0.0, 0.0), 2),
                         DualNum.constant(0.0, 2)
                 );
+
+                pose = pose.plus(delta.value());
+                return delta.velocity().value();
```
Applying a zero delta to `pose` shouldn't change it so you can just return zero velocity here.

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -166,14 +177,17 @@ public Twist2dDual<Time> update() {
 
             double headingDelta = heading.minus(lastHeading);
-            Twist2dDual<Time> twist = kinematics.forward(new MecanumKinematics.WheelIncrements<>(
+            Twist2dDual<Time> twist = MecanumDrive.this.kinematics.forward(new MecanumKinematics.WheelIncrements<>(
```
Ditto - can this just be `kinematics.forward(...`?

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -199,16 +213,18 @@ public Twist2dDual<Time> update() {
             lastHeading = heading;
 
-            return new Twist2dDual<>(
+            delta = new Twist2dDual<>(
```
Making the `delta` is kind of unnecessary here because it's immediately split. You should be able to pass `Twist2d(twist.line.value(), headingDelta)` directly to `pose.plus()` and then return `twist.velocity().value()`.

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -440,18 +456,14 @@ public void preview(Canvas c) {
-        }
-
-        estimatedPoseWriter.write(new PoseMessage(pose));
+    public Pose2d getLocalizerPose() {
+        return localizer.getPose();
+    }
```
`localizer.getPose()` is about the same length as `getLocalizerPose()`... I prefer the former.

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -440,18 +456,14 @@ public void preview(Canvas c) {
-        pose = pose.plus(twist.value());
-
-        poseHistory.add(pose);
-        while (poseHistory.size() > 100) {
-            poseHistory.removeFirst();
-        }
-
-        estimatedPoseWriter.write(new PoseMessage(pose));
```
Why did we get rid of this?

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TankDrive.java:
```diff
@@ -180,26 +192,29 @@ public Twist2dDual<Time> update() {
             meanRightVel /= rightEncs.size();
 
             FlightRecorder.write("TANK_LOCALIZER_INPUTS",
-                     new TankLocalizerInputsMessage(leftReadings, rightReadings));
```
nit: add back the space

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/ThreeDeadWheelLocalizer.java:
```diff
@@ -95,6 +115,9 @@ public Twist2dDual<Time> update() {
         lastPerpPos = perpPosVel.position;
 
-        return twist;
+        delta = twist;
+
+        pose = pose.plus(delta.value());
+        return delta.velocity().value();
```
Can you just use `twist`?

rbrott commented:
> I did this (I added a `getLocalizerPose` method to the drive classes too, but the drive class no longer stores a pose independently of the localizers), but doing this would cause a lot of issues with people updating. no?

It will cause some issues. Every quickstart change is painful to some degree. Even changes that simply add behavior are annoying for teams to pull in because you can't just increment a version number. My only solution is to push as much code as possible out of the quickstart itself and be deliberate about which changes are merged.

zachwaffle4 commented:
> My only solution is to push as much code as possible out of the quickstart itself and be deliberate about which changes are merged.

Could localization (and maybe even the drive classes themselves or some kind of interface defining drive class behavior) be moved to roadrunner-ftc, then?

rbrott commented:
> Could localization (and maybe even the drive classes themselves or some kind of interface defining drive class behavior) be moved to roadrunner-ftc, then?

It's hard for me to imagine further changes I'll want to make to the drive and dead wheel classes, and I'm pretty happy with how they're implemented. I never liked the 0.x RR API of extending `MecanumDrive` in the quickstart. Inheritance is bad even when the API is designed for it. I also think it's good for teams to have some exposure to how things are implemented.

I do think `Localizer` itself could maybe be moved to RR FTC. I anticipate adding OTOS/Pinpoint drivers at some point, and I suspect it may take a few tries to get them working exactly how I would like (e.g., plumbing faults through, estimating velocity). It's also not super important for teams to understand how the driver works. Having the localizer interface defined in RR FTC would also make it easier for third parties to use it.

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -199,16 +205,16 @@ public Twist2dDual<Time> update() {
             lastHeading = heading;
 
-            return new Twist2dDual<>(
+            pose = pose.plus(new Twist2dDual<>(
```
I think you can just do `pose = pose.plus(new Twist2d(twist.line.value(), headingDelta))`

zachwaffle4 replied:
Does it matter? (genuine question)

rbrott replied:
Yes? It's like having `Vector2d(x, y).x + 2` instead of `x + 2` to me.

rbrott commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java:
```diff
@@ -439,19 +445,19 @@ public void preview(Canvas c) {
         }
     }
-
+    
```
nit: remove tab

zachwaffle4 commented:
> I also think it's good for teams to have some exposure to how things are implemented.

I think it's just easier for teams to update when there are fewer files in the quickstart (also I dislike the idea of quickstarts but that's a conversation for another day); for example, my teamcode repo is organized in a way where all of the quickstart files are in a `roadrunner` subpackage so I can't just git pull in changes should this quickstart (or Jay's otos version) update.

Other than that conversation I believe all of the change requests have finally been resolved!

rbrott reviewed (approved):
I concur! Thanks for bearing with me
